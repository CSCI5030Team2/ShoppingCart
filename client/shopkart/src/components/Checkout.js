import React, {Component} from "react";
import { connect } from "react-redux";
import {putCheckout} from "../actions/products"
import LoginNavbar from "./LoginNavbar"
import axios from "axios"

const initialState = {
    cardNumber: "",
    expiryMonth:"",
    expiryYear:"",
    cvv:"",
    cardHolder:"",
    cardNumberError: "",
    expiryMonthError: "",
    expiryYearError: "",
    cvvError:"",
    cardHolderError:""
};

export class Checkout extends Component {
    constructor(props) {
        super(props);
        this.onPayment = this.onPayment.bind(this);
    }
    state = {
        initialState
    };


    onChange = event => {
        this.setState({ [event.target.name]: event.target.value }); 
    };

    //validation of each input in card details
    validate = () => {
        let cardNumberError = "";
        let expiryMonthError = "";
        let expiryYearError = "";
        let cvvError = "";
        let cardHolderError = "";
        if(this.state.cardNumber) {
            cardNumberError = "Card number required"
        }
        if(!this.state.expiry) {
            expiryMonthError = "Expiry required"
        }
        if(!this.state.expiry) {
            expiryYearError = "Expiry required"
        }
        if(!this.state.cvv) {
            cvvError = "CVV required"
        }
        if(!this.state.cardHolder) {
            cardHolderError = "Card Holder Name required"
        }
        if (cardNumberError || expiryMonthError || expiryYearError || cvvError || cardHolderError ) {
            this.setState({cardHolderError:cardHolderError, expiryMonthError:expiryMonthError, expiryYearError:expiryYearError, cvvError:cvvError, cardHolderError:cardHolderError});

            return false;
        }
        return true;
    };

    //function to handle changes during submit
    handleSubmit = event => {
        event.preventDefault();
        const isValid = this.componentDidCatch.validate();
        if(isValid){
            console.log(this.state)
            this.setState(initialState)
        }
    }

    //funtion containing logic of payment
    onPayment = e => {
        e.preventDefault();
        const isValid = this.validate();
        if(isValid) {
            let cardDetails = {
                cardNumber: this.state.cardNumber,
                expiry: this.state.expiry,
                cvv: this.state.cvv,
                cardHolder: this.state.cardHolder
            };
            this.props.putCheckout(cardDetails);
            this.setState({
                cardNumber:"",
                expiry:"",
                cvv:"",
                cardHolder:""
            })
            alert("Payment successful");
        }
    }

    //BootStrap code for input boxes and submit button
    render() {
        console.log(this.props);
            return(
                    <div>
                        <LoginNavbar />
                    <div
                        className="container"
                        align="center"
                        style={{
                            width: 20 + "em",
                            marginTop: 3 + "em",
                            height: "auto"
                        }}
                    >
                    <form>
                                <h1>Payment Page</h1>
                                <p>
{/* enter card number, expiry month, expiry year, cvv, card holder name to make payment using credit card */}
                                    <input
                                        type="text"
                                        name="cardNumber"
                                        placeholder="Enter Card Number"
                                        maxLength={16}
                                        value={this.state.cardNumber}
                                        onChange= {this.onChange}
                                        required
                                    />
                                </p>
                                <p>
                                    <input
                                        type="text"
                                        name="exipryMonth"
                                        placeholder="Month MM"
                                        maxLength={2}
                                        value={this.state.expiryMonth}
                                        onChange= {this.onChange}
                                        required
                                    
                                    />
                                </p>
                                <p>
                                    <input
                                        type="text"
                                        name="expiryYear"
                                        placeholder="Year YYYY"
                                        maxLength={4}
                                        value={this.state.expiryYear}
                                        onChange= {this.onChange}
                                        required
                                    
                                    />
                                </p>
                                <p>
                                    <input
                                        type="text"
                                        name="cvv"
                                        placeholder="Enter CVV"
                                        maxLength={3}
                                        value={this.state.cvv}
                                        onChange= {this.onChange}
                                        required
                                    
                                    />
                                </p>
                                <p>
                                    <input
                                        type="text"
                                        name="name"
                                        placeholder="Enter Card Holder Name"
                                        value={this.state.cardHolder}
                                        onChange= {this.onChange}
                                        required
                                    
                                    />
                                </p>
                                <p>
                                    <div className="checkouterror" style={{ fontSize: 15, color: "red" }}>
                                    {/* dispatch error from node -vivek */}
                                    {this.props.error ? (
                                        <>
                                        <p>{this.props.error}</p>
                                        </>
                                    ) : (
                                        <p>{this.state.cardNumberError||this.state.expiryonthError||this.state.expiryYearError||this.state.cvvError||this.state.cardHolderError}</p>
                                    )}
                                    </div>
                                </p>

                                <button
                                    onChange={this.handleChange}
                                    onClick={() => {

                                        //sending token to backend route. The token is stored in localStorage
                                        axios.put("http://localhost:8080/cart",
                                            {
                                                token: localStorage.getItem("token")
                                            }

                                        ).then(r => console.log(r.data))

                                        //alert message after success
                                        alert("PAYMENT SUCCESS")
                                    }}
                                    >
                                    Payment
                                </button>
                            </form>
                        </div>
                        </div>
            )
    }
}

const mapStateToProps = state => ({
    products: state.productReducer.products
});

export default connect(
    mapStateToProps,
    { putCheckout}
)(Checkout);