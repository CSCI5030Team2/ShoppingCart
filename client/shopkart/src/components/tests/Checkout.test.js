import Checkout from "../Checkout";
import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import store from "../../../src/store";
import { BrowserRouter } from "react-router-dom";


const checkout = jest.fn();
const wrapper = mount(
  <Provider store={store}>
  <BrowserRouter>
    <Checkout checkout={checkout} />
    </BrowserRouter>
  </Provider>
);

describe("Test Checkout Component"), () => {
    beforeAll(() => {
        jest.spyOn(console, 'log').mockImplementation(() => {});
    });

afterAll(() => {
    console.log.mockRestore();
});

afterEach(() => {
    console.log.mockClear();
});

it("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
});

it("should have exactly five input fields", () => {
    expect(wrapper.find("input").length).toBe(5);
});

it("should have one form component", () => {
    expect(wrapper.find("form").length).toBe(1);
});

  it("should have exactly one button", () => {
    expect(wrapper.find("button").length).toBe(1);
});

it("should have Enter Card Number placeholder on CardNumber input field", () => {
    expect(
      wrapper
        .find("input")
        .at(0)
        .props().placeholder
    ).toBe("Enter Card Number");
  });

  it("should have Month MM placeholder on ExpiryMonth input field", () => {
    expect(
      wrapper
        .find("input")
        .at(1)
        .props().placeholder
    ).toBe("Month MM");
  });

  it("should have Year YYYY placeholder on ExpireYear input field", () => {
    expect(
      wrapper
        .find("input")
        .at(2)
        .props().placeholder
    ).toBe("Year YYYY");
  });


  it("should have Enter CVV placeholder on CVV input field", () => {
    expect(
      wrapper
        .find("input")
        .at(3)
        .props().placeholder
    ).toBe("Enter CVV");
  });

  it("should have Enter Card Holder Name placeholder on CardHolderName input field", () => {
    expect(
      wrapper
        .find("input")
        .at(4)
        .props().placeholder
    ).toBe("Enter Card Holder Name");
  });
}