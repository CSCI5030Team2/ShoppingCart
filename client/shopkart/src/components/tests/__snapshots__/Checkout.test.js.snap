// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Checkout Component should render the component 1`] = `ReactWrapper {}`;
